cryptonight/src/hash_v4.rs

```diff
@@ -0,0 +1,566 @@
+    // faster than Intel Sandy Bridge and newer processors Surprisingly, Intel
+    // Nehalem also has 1-cycle ROR/ROL, so it'll also be faster than Intel Sandy
+    // Bridge and newer processors AMD Bulldozer has 4 cycles latency for MUL
+    // (slower than Intel) and 1 cycle for ROR/ROL (faster than Intel), so average
+    // performance will be the same Source: https://www.agner.org/optimize/instruction_tables.pdf
+    const OP_LATENCY: [usize; INSTRUCTION_COUNT] = [3, 2, 1, 2, 2, 1];
+
+    // Instruction latencies for theoretical ASIC implementation
+    const ASIC_OP_LATENCY: [usize; INSTRUCTION_COUNT] = [3, 1, 1, 1, 1, 1];
+
+    // Available ALUs for each instruction
+    const OP_ALUS: [usize; INSTRUCTION_COUNT] = [
+        ALU_COUNT_MUL,
+        ALU_COUNT,
+        ALU_COUNT,
+        ALU_COUNT,
+        ALU_COUNT,
+        ALU_COUNT,
+    ];
+
+    let mut data = [0_u8; 32];
+    data[0..8].copy_from_slice(&height.to_le_bytes());
+
+    data[20] = -38_i8 as u8; // change seed
+
+    // Set data_index past the last byte in data
+    // to trigger full data update with blake hash
+    // before we start using it
+    let mut data_index: usize = data.len();
+
+    let mut code_size: usize;
+
+    // There is a small chance (1.8%) that register R8 won't be used in the
+    // generated program, so we keep track of it and try again if it's not used
+    loop {
+        let mut latency = [0_usize; 9];
+        let mut asic_latency = [0_usize; 9];
+
+        // Tracks previous instruction and value of the source operand for
+        // registers R0-R3 throughout code execution:
+        // byte 0: current value of the destination register
+        // byte 1: instruction opcode
+        // byte 2: current value of the source register
+        //
+        // Registers R4-R8 are constant and are treated as having the same
+        // value, because when we do the same operation twice with two constant
+        // source registers, it can be optimized into a single operation.
+        let mut inst_data: [usize; 9] =
+            [0, 1, 2, 3, 0xFFFFFF, 0xFFFFFF, 0xFFFFFF, 0xFFFFFF, 0xFFFFFF];
+
+        let mut alu_busy = [[false; ALU_COUNT]; TOTAL_LATENCY + 1];
+        let mut is_rotation = [false; INSTRUCTION_COUNT];
+        is_rotation[Ror as usize] = true;
+        is_rotation[Rol as usize] = true;
+        let mut rotated = [false; 4];
+        let mut rotate_count = 0_usize;
+
+        let mut num_retries = 0_usize;
+        code_size = 0;
+
+        let mut total_iterations = 0_usize;
+        let mut r8_used = false;
+
+        // Generate random code to achieve minimal required latency for our abstract CPU
+        // Try to get this latency for all 4 registers
+        while ((latency[0] < TOTAL_LATENCY)
+            || (latency[1] < TOTAL_LATENCY)
+            || (latency[2] < TOTAL_LATENCY)
+            || (latency[3] < TOTAL_LATENCY))
+            && (num_retries < 64)
```

Comment:
Extra parentheses aren't needed in Rust, only needed for separating operations.
```suggestion
        while (latency[0] < TOTAL_LATENCY
            || latency[1] < TOTAL_LATENCY
            || latency[2] < TOTAL_LATENCY
            || latency[3] < TOTAL_LATENCY)
            && num_retries < 64
```